source/server/connection_handler_impl.cc

```diff
@@ -52,133 +56,142 @@ void ConnectionHandlerImpl::addListener(absl::optional<uint64_t> overridden_list
-    details->listener_ = std::move(udp_listener);
-  }
-
-  if (disable_listeners_) {
-    details->listener_->pauseListening();
-  }
-  if (auto* listener = details->listener_->listener(); listener != nullptr) {
-    listener->setRejectFraction(listener_reject_fraction_);
+    for (auto& socket_factory : config.listenSocketFactories()) {
+      details->addActiveListener(
+          config, socket_factory, listener_reject_fraction_, disable_listeners_,
+          config.udpListenerConfig()->listenerFactory().createActiveUdpListener(
+              runtime, *worker_index_, *this,
+              config.listenSocketFactories()[0]->getListenSocket(*worker_index_), dispatcher_,
+              config));
+    }
   }
 
-  details->listener_tag_ = config.listenerTag();
-  details->address_ = config.listenSocketFactory().localAddress();
-
   ASSERT(!listener_map_by_tag_.contains(config.listenerTag()));
 
-  listener_map_by_tag_.emplace(config.listenerTag(), details);
-  // This map only store the new listener.
-  if (absl::holds_alternative<std::reference_wrapper<ActiveTcpListener>>(
-          details->typed_listener_)) {
-    tcp_listener_map_by_address_.insert_or_assign(
-        config.listenSocketFactory().localAddress()->asStringView(), details);
-
-    auto& address = details->address_;
-    // If the address is Ipv6 and isn't v6only, parse out the ipv4 compatible address from the Ipv6
-    // address and put an item to the map. Then this allows the `getBalancedHandlerByAddress`
-    // can match the Ipv4 request to Ipv4-mapped address also.
-    if (address->type() == Network::Address::Type::Ip &&
-        address->ip()->version() == Network::Address::IpVersion::v6 &&
-        !address->ip()->ipv6()->v6only()) {
-      if (address->ip()->isAnyAddress()) {
-        // Since both "::" with ipv4_compat and "0.0.0.0" can be supported.
-        // If there already one and it isn't shutdown for compatible addr,
-        // then won't insert a new one.
-        auto ipv4_any_address = Network::Address::Ipv4Instance(address->ip()->port()).asString();
-        auto ipv4_any_listener = tcp_listener_map_by_address_.find(ipv4_any_address);
-        if (ipv4_any_listener == tcp_listener_map_by_address_.end() ||
-            ipv4_any_listener->second->listener_->listener() == nullptr) {
-          tcp_listener_map_by_address_.insert_or_assign(ipv4_any_address, details);
-        }
-      } else {
-        auto v4_compatible_addr = address->ip()->ipv6()->v4CompatibleAddress();
-        // Remove this check when runtime flag
-        // `envoy.reloadable_features.strict_check_on_ipv4_compat` deprecated.
-        // If this isn't a valid Ipv4-mapped address, then do nothing.
-        if (v4_compatible_addr != nullptr) {
-          tcp_listener_map_by_address_.insert_or_assign(v4_compatible_addr->asStringView(),
-                                                        details);
+  for (auto& per_address_details : details->per_address_details_list_) {
```

Comment:
I may have missed it, but is there a test that adds more than one listener?
Maybe this is planned as part of the next PRs.

Reply:
Sorry, I forget to cherry-pick that commit. Let me append that to this PR